security/src/main/java/io/airlift/security/utils/CloseableGSSContext.java

```diff
@@ -0,0 +1,180 @@
+package io.airlift.security.utils;
+
+import com.google.common.base.Throwables;
+import com.sun.security.auth.module.Krb5LoginModule;
+import io.airlift.log.Logger;
+import io.airlift.security.exception.AuthenticationException;
+import org.ietf.jgss.GSSContext;
+import org.ietf.jgss.GSSCredential;
+import org.ietf.jgss.GSSException;
+import org.ietf.jgss.GSSManager;
+import org.ietf.jgss.GSSName;
+import org.ietf.jgss.Oid;
+
+import javax.security.auth.Subject;
+import javax.security.auth.login.AppConfigurationEntry;
+import javax.security.auth.login.Configuration;
+import javax.security.auth.login.LoginContext;
+import javax.security.auth.login.LoginException;
+
+import java.io.Closeable;
+import java.net.InetAddress;
+import java.net.URI;
+import java.net.UnknownHostException;
+import java.security.Principal;
+import java.security.PrivilegedAction;
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+import static javax.security.auth.login.AppConfigurationEntry.LoginModuleControlFlag.REQUIRED;
+
+public class CloseableGSSContext
+        implements Closeable
+{
+    private static final Logger log = Logger.get(KerberosUtil.class);
+
+    private final GSSContext context;
+    private final LoginContext loginContext;
+
+    public CloseableGSSContext(String serviceName, URI serviceUri)
+    {
+        requireNonNull(serviceName, "serviceName is null");
+        requireNonNull(serviceUri, "serviceUri is null");
+        try {
+            loginContext = new LoginContext("", null, null, new Configuration()
+            {
+                @Override
+                public AppConfigurationEntry[] getAppConfigurationEntry(String name)
+                {
+                    Map<String, String> options = new HashMap<>();
```

Comment:
@martint / @dain / @electrum , quick q. Is presto-client used by any service? If it is, we need to allow people to set client principal as well. Thanks!